rfcs/20200612-stream-executor-c-api/C_API_versioning_strategy.md

```diff
@@ -0,0 +1,226 @@
+
+```cpp
+// TF Implementation
+
+void TF_Foo(const SE_Device* device) {
+  // TF checks for struct_size greater than 32.
+  if (device->struct_size > offsetof(SE_Device, device_handle)) {
+     // TF knows that device_handle can be safely read from.
+     DoSomething(device->device_handle);
+  }
+}
+```
+
+### Forwards compatibility
+ 
+In the event that a plugin is up to date or newer, se->struct_size would have been initialized to 48. This would then pass the TF_Foo() check above and device_handle can be safely accessed.
+ 
+**Future Plugin compiled against StreamExecutorInterface v3**
+
+```cpp
+// StreamExecutorInterface header version 3
+typedef struct SE_Device {
+  size_t struct_size;
+  void* next;
+
+  const char* name;
+  size_t name_len;
+  void* device_handle;
+  void* data;          // Added in v3
+} SE_Device;
+
+// Evaluates to 48
+#define SE_DEVICE_STRUCT_SIZE TF_OFFSET_OF_END(SE_Device, data)
+
+
+// Plugin Implementation
+
+SE_Device* Plugin_CreateDevice() {
+  SE_Device* se = new SE_Device{ SE_DEVICE_STRUCT_SIZE };
+  // se->struct_size will be 48
+  ...
+  return se;
+}
+```
+ 
+Using the same TF_Foo() above, TF_Foo()  was implemented before SE_Device::data was added after SE_Device::device_handle. Since TensorFlow only knows about the members that come before SE_Device::data, the newly added device->data will not be accessed.
+ 
+## When TensorFlow deprecates functionality
+When functionality is being deprecated, there will be comments next to the member indicating so. The member is left in place to preserve the alignment and offset of the existing structure members.
+ 
+Since members are not allowed to be removed or reordered, refactors (e.g. renaming device_handle to dev_handle) or changing of member types (e.g. from int to float) are considered as deprecation.
+ 
+### Backwards compatibility
+SE_Device::data has been deprecated in version 4, and a comment in the header indicated as such.
+ 
+**Future TensorFlow compiled against StreamExecutorInterface v4**
+
+```cpp
+// StreamExecutorInterface header version 4
+typedef struct SE_Device {
+  size_t struct_size;
+  void* next;
+
+  const char* name;
+  size_t name_len;
+  void* device_handle;
+  void* data;            // Deprecated
+} SE_Device;
+
+// Evaluates to 48
+#define SE_DEVICE_STRUCT_SIZE TF_OFFSET_OF_END(SE_Device, data)
+
+// TF Implementation
+
+void TF_Foo(const SE_Device* device) {
+  // TF checks for struct_size greater than 32.
+  if (device->struct_size > offsetof(SE_Device, device_handle)) {
+     // TF knows that device_handle can be safely accessed.
+     DoSomething(device->device_handle);
+  }
+
+  // TensorFlow removes implementation to stop using deprecated functionality.
+  /*
+  if (device->struct_size > offsetof(SE_Device, data)) {
+     // TF knows that device->data can be safely accessed.
+     DoSomething(device->data);
+  }
+  */
+
+}
+``
```

Comment:
Missing backtick.

Reply:
Fixed in https://github.com/tensorflow/community/pull/257/commits/d54e1ee60a77915213c24cb03264f82486d1d09d.